### PR TITLE
feat: Added HasGameReference mixin

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

This almost exactly like the current `HasGameRef` mixin, except that:
 - The property is called `game` instead of `gameRef` (the "gameRef" violates Dart naming conventions against using abbreviations in variable names);
 - The template type `T` supports all `Game`s, not only `FlameGame`s;
 - Better integration with the `SingleGameInstance` mixin;
 - The new mixin is within `experimental`, to reduce chance that it will confuse the users.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [-] I have updated/added relevant examples in `examples`.

## Breaking Change

- [-] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

## Related Issues

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Review by spydon — APPROVED** [review]
Should `HasGameRef` also be deprecated?

**Review by erickzanardo — CHANGES_REQUESTED** [review]
I don't like the idea of having two things that do practically the same thing only with a few differences.

The alternative here would be to deprecate HasGameRef or integrate the changes that were introduced on this new mixin there.

To avoid a breaking change I would prefer to just integrate the generics and better integration on the single game mixin on the has game ref

**Comment by st-pasha** [comment]
@erickzanardo Shouldn't we take a more careful approach here: instead of merging the two classes, merely mark the old one as deprecated? That way the users will be able to switch at their own pace, instead of being forced into it.

**Comment by erickzanardo** [comment]
@st-pasha from what I see on your implementation, despite the name changes, everything could be accommodated on the current `HasGameRef` implementation.

The generics would change, but a `FlameGame` is a `Game`, so there wouldn't be a breaking change here. And how you made the retrievement of the game instance is an internal thing, just like on the `HasGameRef`, so that could go in there too.

I know that the naming convention may be off, but I would rather keep it instead of adding a big (big because this is widely used) just because of renaming, and avoid frutastions like [this](https://github.com/flame-engine/flame/discussions/1401#discussioncomment-3212401).

**Comment by st-pasha** [comment]
Ok, I've merged the two implementations then: the old name is marked as "Deprecated", but only in comments (so everything continues to work as before). We could upgrade the comment to "real" deprecate-annotation, either in this PR or at some later time (perhaps after the next release).

**Comment by renancaraujo** [comment]
Sorry for popping in so late into the discussion, but I seriously think that we should slow down on breaking core API. I agree that `HasGameReference` is better than `HasGameRef` and a `game` property makes more sense than `gameRef`, **but**, renaming the core API has some serious trade-offs for example: making existing tutorial blogs and videos totally outdated, let alone all the projects people have already deployed 

In my opinion, the juice doesn't worth the squeeze here. The rule of thumb on changing core API should be "we rename it only if it becomes misleading or sufficiently inaccurate". 

All the other aspects of this PR are great and I appreciate all the work done here ❤️

**Comment by st-pasha** [comment]
@renancaraujo I agree completely, breaking changes should be avoided at all costs, and when necessary introduced at a very slow pace. Which is why in the first submission I had the `HasGameReference` mixin completely separate from `HasGameRef`. This way we can transition from the old mixin to the new one reaaaally slow:
 - at first start mentioning in Discord that we prefer `HasGameReference` mixin now;
 - change our documentation/examples to use the new mixin only;
 - add a note somewhere that `HasGameRef` will eventually be deprecated;
 - deprecate it only at some future date, say in 6 months or in a year;
 - finally remove after it was deprecated for some long time.

This way we could still introduce code improvements, but with no disruption to the users.

The only con of this approach is that it results in some code duplication -- but I don't see it as that much of a big deal. Code duplication when migrating between different APIs is quite common, and doesn't result in increased maintenance cost.

Should I revert to that original version of the PR?

**Comment by spydon** [comment]
I agree with @renancaraujo and @erickzanardo, it seems unnecessary to rename the mixin and `gameRef` -> `game` just because it breaks the naming convention. The PR overall is good though, can't the `HasGameRef` just be changed to have the behaviour of `HasGameReference`, without changing any naming?

**Comment by st-pasha** [comment]
Alright, I've restored `HasGameRef` to the way it originally was. Now:

1. No changes for existing users whatsoever (not even deprecation);
2. If anyone gets annoyed by less-than-perfect grammar (like me), they can switch to `HasGameReference` and `.game`.

Thus, this PR is a net-positive for some users, and neutral for everyone else.

**Comment by spydon** [comment]
> Alright, I've restored `HasGameRef` to the way it originally was. Now:
> 
>     1. No changes for existing users whatsoever (not even deprecation);
> 
>     2. If anyone gets annoyed by less-than-perfect grammar (like me), they can switch to `HasGameReference` and `.game`.
> 
> 
> Thus, this PR is a net-positive for some users, and neutral for everyone else.

I don't think that we should keep `HasGameReference`, it will just lead to confusion having two mixins doing the same thing.

**Comment by spydon** [comment]
> But that is exactly how code evolution works, if we want to avoid breaking changes.

This is not a necessary evolution though and we can avoid it. The code changes can be done in `HasGameRef` and the name change is not more important than avoiding a breaking change even though it would be a slow transition, the whole team has agreed on this.

**Comment by st-pasha** [comment]
> This is not a necessary evolution though and we can avoid it. The code changes can be done in HasGameRef and the name change is not more important than avoiding a breaking change even though it would be a slow transition

So, what you're saying is that we want to avoid even a slow deprecation cycle here?

It's fine by me -- in fact, the PR is already structured that way. We could have the two mixins live side-by-side, giving the users an opportunity to choose which one they prefer. Perhaps I could add some more clarification in doc-comments about this, so as to avoid any potential confusion on the users part.

**Comment by spydon** [comment]
> So, what you're saying is that we want to avoid even a slow deprecation cycle here?

No, what I'm saying is that the `HasGameReference` mixin shouldn't be introduced at all.
The changes done to the code in `HasGameReference` should be possible to add in a non-breaking way to `HasGameRef`, right?

**Comment by st-pasha** [comment]
> No, what I'm saying is that the HasGameReference mixin shouldn't be introduced at all.

In that case I'm not exactly sure what is the objection against having this mixin? Given that it has some benefit (which appears minor for some people, and substantial for others), and no cost (anyone who has a question why there are two mixins can read the doc-comments and find the answer).

**Comment by erickzanardo** [comment]
Having `HasGameReference` means that we would deprecate `HasGameRef` eventually, which we won't because of the reasons mentioned here, specially because of this one mentioned by Renan:

> renaming the core API has some serious trade-offs for example: making existing tutorial blogs and videos totally outdated, let alone all the projects people have already deployed

Maybe this is something that we can consider rename when we start planning a Flame 2.0. But that is something for the long term future.

**Comment by spydon** [comment]
> In that case I'm not exactly sure what is the objection against having this mixin? Given that it has some benefit (which appears minor for some people, and substantial for others), and no cost (anyone who has a question why there are two mixins can read the doc-comments and find the answer).

I don't understand, what are the benefits that can't be incorporated into the current mixin (except for the name changes)?
We shouldn't have two ways of doing things if we can avoid it, even if we can write docs about it, users are generally bad at reading the docs well.

> Maybe this is something that we can consider rename when we start planning a Flame 2.0. But that is something for the long term future.

Agree.

**Comment by st-pasha** [comment]
I'll try to summarize the state of the discussion so far, so that it is easier to move forward:

1. I believe we all agree that, by itself, and not taking other factors into account, the name change `gameRef` -> `game` is good. There is a difference in opinion of how much of a benefit this creates, but at least no one had said that the old name was better.

2. We also definitely agree that breaking changes ought to be avoided, especially if done for minor reasons like changing a slightly imperfect name.
    - We do recognize that the current PR does not create any kind of breaking change. We want it to stay that way.

The only point of disagreement is whether there will be a substantial amount of confusion, and whether the drawbacks of that confusion outweigh the benefits of having a better name. I'm not sure how to quantify this confusion, but as far as I can see, the confusion is:
- **harmless**, in the sense that even if someone picks one of the two mixins at random, their code would still work the same (up to a name change);
- **resolvable**, meaning that if someone does see two mixins with similar names and the question of why there are two classes really bothers them, they would just go and check the documentation which would state the reasons plainly;
- **non-unique**: there are other places in the code base where we are doing a transition from old way of doing things to the new way, this is not a "bug", this is the way code evolution works. If the users can understand this principle, it would make their life easier.

**Comment by spydon** [comment]
> The only point of disagreement is whether there will be a substantial amount of confusion, and whether the drawbacks of that confusion outweigh the benefits of having a better name. I'm not sure how to quantify this confusion, but as far as I can see, the confusion is:
> 
>     * **harmless**, in the sense that even if someone picks one of the two mixins at random, their code would still work the same (up to a name change);

It won't work the same if they choose the wrong one when they are following a tutorial, due to the name changes.

>     * **resolvable**, meaning that if someone does see two mixins with similar names and the question of why there are two classes really bothers them, they would just go and check the documentation which would state the reasons plainly;

They don't necessarily see that there are two classes, and just pick the wrong one and it won't work for them. 

>     * **non-unique**: there are other places in the code base where we are doing a transition from old way of doing things to the new way, this is not a "bug", this is the way code evolution works. If the users can understand this principle, it would make their life easier.

We do recognize this, but this small change just isn't worth it since it is simply a change of naming (and the class is a very commonly used one), the whole team agrees on this.

Just like Erick said, this change can be something to be considered for v2.0, but it won't be merged with the name change before that (the code change is fine however, so that could just be put into `HasGameRef` instead).

File: packages/flame/test/components/mixins/has_game_reference_test.dart [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        final component3 = _Component()..addToParent(component2);
+        expect(component3.game, game);
+      },
+    );
+  });
+}
+
+class _Component<T extends Game> extends Component with HasGameReference<T> {}
+
+class _MyGame extends FlameGame {}
```
Comment (alestiago) [inline packages/flame/test/components/mixins/has_game_reference_test.dart:89]: Nit: can we move this at the start of the file?

As a reader I prefer reading the class definitions before I encounter them. I believe uncle Bob mentions something related to this in his Clean Code Book.

**Comment by alestiago** [comment]
> I believe we all agree that, by itself, and not taking other factors into account, the name change gameRef -> game is good. There is a difference in opinion of how much of a benefit this creates, but at least no one had said that the old name was better.

Fully agree!

File: packages/flame/lib/src/components/mixins/has_game_ref.dart [hunk 2]
```diff
@@ -2,6 +2,7 @@ import 'package:flame/components.dart';
 import 'package:flame/game.dart';
 import 'package:flutter/material.dart';
 
+/// See also: `HasGameReference` mixin
 mixin HasGameRef<T extends FlameGame> on Component {
```
Comment (alestiago) [inline packages/flame/lib/src/components/mixins/has_game_ref.dart:6]: Can this be now a `typedef`?

**Comment by spydon** [comment]
> Fully agree!
> 
> I'm in favour of the rename and fully agree with @st-pasha . The current naming is not great since it violates effective dart (https://dart.dev/guides/language/effective-dart/design#avoid-abbreviations).
> 
> I'm all in in deprecating HasGameRef for HasGameReference.

It will be deprecated, but not pre-v2. This discussion is already settled.
We can create a ticket for it (with a v2 label) so that we don't forget about it. 

What's left to do for this PR to get merged is to move the new code into the old mixin in a non-breaking way, could you do that @st-pasha?

File: packages/flame/test/components/mixins/has_game_reference_test.dart (hunk 1 above)
Comment (st-pasha) [reply packages/flame/test/components/mixins/has_game_reference_test.dart:89]: The number one reason people visit a test file is because some tests are failing and you need to figure out what's wrong. So, with that purpose in mind, you'd open a test file - and press the green arrow next to the `void main()`. The IDE then runs all tests showing the failing one(s), and only then you'd jump there to study those tests in more detail.

Thus, when it comes to test files, it is more practical to have the "main" function at the top. All the other definitions can live at the bottom - after all, you don't want to read all of them, only those that are relevant to the test that is failing.

**Comment by alestiago** [comment]
It is very nice that this change will get in! Great news 🔥💙

Thanks @st-pasha for working on it!

**Comment by kurtome** [comment]
> It will be deprecated, but not pre-v2.

No one asked me, but this is the way. Tracking breaking changes and targeting them for v2 (possible years out, I assume) is definitely the ideal for a project with this many users

**Comment by spydon** [comment]
> No one asked me, but this is the way. Tracking breaking changes and targeting them for v2 (possible years out, I assume) is definitely the ideal for a project with this many users

Your opinion is always valued @kurtome, and that's a good point actually, we need to start tracking what we want to put in V2.

**Comment by st-pasha** [comment]
> we need to start tracking what we want to put in V2.

Perhaps we could have an issue with a task list of things, and then when the time comes we could quickly convert those task items into issues with a press of a button.

**Comment by st-pasha** [comment]
@spydon I've merged `HasGameReference` into the `HasGameRef`.

File: packages/flame/lib/src/components/input/hud_button_component.dart [hunk 3]
```diff
@@ -8,7 +9,7 @@ import 'package:flutter/rendering.dart' show EdgeInsets;
 /// Note: You have to set the [button] in [onLoad] if you are not passing it in
 /// through the constructor.
 class HudButtonComponent extends ButtonComponent
-    with HasGameRef, ComponentViewportMargin {
+    with HasGameRef<FlameGame>, ComponentViewportMargin {
```
Comment (spydon) [inline packages/flame/lib/src/components/input/hud_button_component.dart:12]: How come we are explicitly adding the generics here now?

File: packages/flame/lib/src/components/mixins/component_viewport_margin.dart [hunk 4]
```diff
@@ -13,7 +13,7 @@ import 'package:meta/meta.dart';
 /// If you set the position of the component instead of a margin when
 /// initializing the component, the margin to the edge of the screen from that
 /// position will be used.
-mixin ComponentViewportMargin on PositionComponent, HasGameRef {
+mixin ComponentViewportMargin on PositionComponent, HasGameRef<FlameGame> {
```
Comment (spydon) [inline packages/flame/lib/src/components/mixins/component_viewport_margin.dart:16]: Same here?

**Review by spydon — APPROVED** [review]
Lgtm, just think that we shouldn't explicitly add the `FlameGame` generics to the `HasGameRef` where it isn't needed.

File: packages/flame/lib/src/components/mixins/has_game_ref.dart [hunk 5]
```diff
@@ -1,41 +1,54 @@
...
 
-mixin HasGameRef<T extends FlameGame> on Component {
-  T? _gameRef;
+/// [HasGameRef] mixin provides property [game] (or [gameRef]), which is the
+/// cached accessor for the top-level game instance.
+///
+/// The type [T] on the mixin is the type of your game class. This type will be
+/// the type of the [game] reference, and the mixin will check at runtime that
+/// the actual type matches the expectation.
+mixin HasGameRef<T extends Game> on Component {
```
Comment (spydon) [inline packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: This is a breaking change changing this from `FlameGame` to `Game`, now I understand why you had to add the explicit generics on the other ones. I understand this is to make this usable by `BareFlameGame`, but is there no other way to do this? Maybe having `BareFlameGame` actually extending `FlameGame` and just throw an `Unimplemented exception` if the camera is accessed?

File: packages/flame/lib/src/components/input/hud_button_component.dart (hunk 3 above)
Comment (st-pasha) [reply packages/flame/lib/src/components/input/hud_button_component.dart:12]: The generic is now based on `Game`, not on `FlameGame`.

Comment (spydon) [reply packages/flame/lib/src/components/input/hud_button_component.dart:12]: Added a comment about it here:
https://github.com/flame-engine/flame/pull/1828#discussion_r979249217

File: packages/flame/lib/src/components/mixins/has_game_ref.dart (hunk 5 above)
Comment (st-pasha) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: I would characterize this as "barely breaking", because in practical usage one would always derive from `FlameGame`:
```dart
class MyGame extends FlameGame with HasTappableComponents, ... {
}
```
and then when using `HasGameRef`, you'd specify your game class:
```dart
class SomeComponent extends PositionComponent with HasGameRef<MyGame> {
}
```

Really, it is only in very corner cases that you'd want to build the game using the plain "FlameGame" class -- such as in simple examples or tests.

Comment (spydon) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: It's very normal to not specify anything there, if you for example want the size or camera etc that you can access directly from the `FlameGame`, so I'd definitely not call this a minor breaking change.

Comment (st-pasha) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: Ok, so should I add the migration instructions? Or revert to the way it was before, with a separate mixin?

Comment (spydon) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: But do we really need it? Can't the `BareFlameGame` just extend `FlameGame` in the manner I described above instead?

Comment (st-pasha) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: That would mean it's not "bare" anymore. We could, in theory, make `FlameGame` derived from `BareFlameGame`, but not the the other way around. 

For example, making `BareFlameGame` derived from `FlameGame` would make it impossible to declare a `camera` property which is incompatible with the FlameGame's camera.

Comment (st-pasha) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: What if we keep `HasGameReference` as it was originally (separate from `HasGameRef`), but inside the `experimental` folder?

Comment (spydon) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: Yeah, we'd just restrict that field and return an unimplemented exception instead, the game would still be bare since it wouldn't use the camera. People can just declare `cameraComponent` if they want to store their camera in a field. I think that is much less messy than putting in this breaking change that breaks lots of tutorials. And I don't think it's a good idea to have both, it's better to wait with those changes until V2.

Comment (st-pasha) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: The whole idea of creating new clean implementations is so that we can leave behind the cruft that accumulated in the old classes. We cannot undermine architectural integrity of the foundational classes in order to get a minor temporary advantage like that.

Comment (spydon) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: On the other hand, it is not that bad if they choose the wrong one here if the only difference is the generics and it won't break current implementations. Alright, I'm onboard with creating `HasGameReference` too in experimental then, and then just have them identical (so the current changes doesn't have to be reverted) except for the generics part.

Comment (spydon) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: @st-pasha did you see my last comment here?

Comment (st-pasha) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart:12]: Yep, just no time right now to work on changes

File: packages/flame/lib/src/components/input/hud_button_component.dart (hunk 3 above)
Comment (st-pasha) [reply packages/flame/lib/src/components/input/hud_button_component.dart]: Reverted this to how it was before

File: packages/flame/lib/src/components/mixins/has_game_ref.dart (hunk 5 above)
Comment (st-pasha) [reply packages/flame/lib/src/components/mixins/has_game_ref.dart]: Done

File: packages/flame/lib/src/components/core/component.dart [hunk 6]
```diff
@@ -375,8 +376,8 @@ class Component {
   ///   - it is invoked when the size of the game canvas is already known.
   ///
   /// If your loading logic requires knowing the size of the game canvas, then
-  /// add `HasGameRef` mixin and then query `gameRef.size` or
-  /// `gameRef.canvasSize`.
+  /// add [HasGameReference] mixin and then query `game.size` or
```
Comment (spydon) [inline packages/flame/lib/src/components/core/component.dart:379]: This should still reference `HasGameRef` since the other one still is in experimental
```suggestion
  /// add [HasGameRef] mixin and then query `game.size` or
```

**Review by spydon — APPROVED** [review]

**Review by erickzanardo — APPROVED** [review]

File: packages/flame/lib/src/components/core/component.dart [hunk 7]
```diff
@@ -6,6 +6,7 @@ import 'package:flame/src/cache/value_cache.dart';
 import 'package:flame/src/components/core/component_set.dart';
 import 'package:flame/src/components/core/position_type.dart';
 import 'package:flame/src/components/mixins/coordinate_transform.dart';
+import 'package:flame/src/experimental/has_game_reference.dart';
```
Comment (spydon) [inline packages/flame/lib/src/components/core/component.dart:9]:
```suggestion
import 'package:flame/src/components/mixins/has_game_ref.dart';
```

**Comment by munsterlander** [comment]
@spydon @st-pasha With this being merged into main, `value_route.dart` in `_sphinx\flame\examples\lib` is throwing errors.  I have synced my fork and ran `pub get` just in case for that yaml file, but I am getting errors like:

```
HasGameReference - Classes can only mix in mixins and classes.
Missing variable type for 'score'.
Try adding an explicit type, or remove implicit-dynamic from your analysis options file.
Undefined name 'game'.
Try correcting the name to one that is defined, or defining the name.
```

Surely I am missing something, right?  As it stands now, docs will not build on any updated branch.

**Comment by spydon** [comment]
> Surely I am missing something, right? As it stands now, docs will not build on any updated branch. Reverting the file to the prior state of [12ce270](https://github.com/flame-engine/flame/commit/12ce270b9b3102b6a9bb1f468369a4fce1e064e6) does solve the issue though.

I don't see any errors regarding this, try to run `melos clean` and then `melos bs`.

**Comment by munsterlander** [comment]
Tried it. I had @ufrshubham check and he had same issue. It only appears once you build the docs locally.

**Comment by munsterlander** [comment]
Solved.  Odd that I was able to get away without using melos for so long until all of a sudden, but none-the-less.
